# 23 - Evaluacion de Modelos de Regresion


* Una vez generado nuestro modelo o hipótesis, tenemos que ser capaces de ***cotejar los resultados obtenidos***


* Para los problemas de regresión la ***medida de calidad*** utilizada es el ***“accuracy”*** y aunque existen varias medidas de accuracy las más empleadas son las siguientes:
<span></span><br><br>
    - ***MAE***: Mean Absolute Error (Error Absoluto Medio)
<span></span><br><br>
    - ***MSE***: Mean Squared Error (Error Cuadrático Medio)
<span></span><br><br>
    - ***Median Absolute error***. Este último es menos utilizado


<hr>


* Veamos a continuación como se calculan esta métricas de calidad y como Scikit las obtiene


* Para ello vamos a utilizar los siguientes resultados siendo:
<span></span><br>
    - ***y_true***: Resultado real
<span></span><br>
    - ***y_predict***: Resultado devuelto por el modelo

In [1]:
y_true = [1, 2, 3]

y_predict = [0.5, 1, 1.5]

<hr>


## MAE: Mean Absolute Error


* Se define como:

<span style="font-size:16px">$$MAE = \frac{1}{n} \sum_{N}^{i=1} \left |  y_i-x_i\right |$$<span>
    
    
* En Scikit lo calcula la función **"mean_absolute_error(y_true, y_predict)"**:

In [2]:
from sklearn.metrics import mean_absolute_error

mean_absolute_error(y_true=y_true, y_pred=y_predict)

1.0

* Veamos como calcularlo:

<span style="font-size:18px">$$\frac{\left | 1-0.5 \right | + \left | 2-1 \right | + \left | 3-1.5 \right |}{3} = 1$$<span>

<hr>


## MSE: Mean Squared Error


* Se define como:

<span style="font-size:16px">$$MSE = \frac{1}{n} \sum_{N}^{i=1} (y_i-x_i)^2$$<span>
    
    
* En Scikit lo calcula la función **"mean_squared_error(y_true, y_predict)"**:

In [3]:
from sklearn.metrics import mean_squared_error

mean_squared_error(y_true=y_true, y_pred=y_predict)

1.1666666666666667

* Veamos como calcularlo:

<span style="font-size:18px">$$\frac{(1-0.5)^2 + (2-1)^2 + (3-1.5)^2}{3} = 1.17$$<span>

<hr>


## Median Absolute error


* Se define como:

<span style="font-size:16px">$$\sum_{N}^{i=1}f_i \geq \frac{n+1}{2}$$<span>
    
    
* En Scikit lo calcula la función **"median_absolute_error(y_true, y_predict)"**:

In [4]:
from sklearn.metrics import median_absolute_error

median_absolute_error(y_true=y_true, y_pred=y_predict)

1.0

* Veamos como calcularlo:

<span style="font-size:18px">$$Mediana[0.5,1,1.5]=1$$<span>

<hr>

*Este Notebook ha sido desarrollado por **Ricardo Moya García** y registrado en Safe Creative como ***Atribución-NoComercial-CompartirIgual***.*

<img src="./imgs/CC_BY-NC-SA.png" alt="CC BY-NC">